config/environments/development.rb

```diff
@@ -54,3 +54,34 @@
   # Allow requests for all domains e.g. <app>.dev.gov.uk
   config.hosts.clear
 end
+
+# make discovery work over HTTP
+module OpenIDConnect
+  module Discovery
+    module Provider
+      class Config
+        class Resource
+          def initialize(uri)
+            @host = uri.host
+            @port = uri.port unless [80, 443].include?(uri.port)
+            @path = File.join uri.path, ".well-known/openid-configuration"
+            @scheme = uri.scheme
+            attr_missing!
```

Comment:
This is interesting, not come across this one before.
Couldn't find it in either the rails docs (found [attribute_missing](https://api.rubyonrails.org/classes/ActiveModel/AttributeMethods.html#method-i-attribute_missing) instead!), nor [ruby](https://ruby-doc.com/search.html?cx=011815814100681837392%3Awnccv6st5qk&q=attr_mising%21&sa=Search&cof=FORID%3A9), nor [this app](https://github.com/alphagov/finder-frontend/search?q=attr_missing)!

I must be missing something obvious, where does this come from and what does it do? (I have my suspicions it might raise an error if one of those uri attributes are missing, but would like to be sure!)

Reply:
It comes from the [`attr_required` gem](https://github.com/nov/attr_required) (which is a dependency of `openid_connect`.  You're right, it raises an exception when an attribute is missing.

Reply:
ah very cool, thanks for the link!